### ISSUE TITLE
Build fails due to go-bindata

### ISSUE THREAD

**Issue by drcrallen** [codedellemc/rexray#850]
# Summary

Build fails due to `go-bindata` problems

# Bug Reports

Build fails with

```
GOOS=linux GOARCH=amd64 go install _/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata
can't load package: package _/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata: cannot find package "_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata" in any of:
        /home/crallen/.go/go-1.8/src/_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata (from $GOROOT)
        /home/crallen/.go/src/src/_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata (from $GOPATH)
Makefile:529: recipe for target '/home/crallen/.go/src/bin/go-bindata' failed
make[1]: *** [/home/crallen/.go/src/bin/go-bindata] Error 1
Makefile:80: recipe for target 'all' failed
make: *** [all] Error 2
```

According to https://github.com/jteeuwen/go-bindata/blob/master/README.md the installation should be 

```
go get -u github.com/jteeuwen/go-bindata/...
```

But after executing that command, the build fails with

```
/home/crallen/.go/src/bin/go-bindata -md5checksum -pkg executors -prefix api/server/executors/bin -o api/server/executors/executors_generated.go api/server/executors/bin/...
flag provided but not defined: -md5checksum
Usage: /home/crallen/.go/src/bin/go-bindata [options] <input directories>

  -debug
        Do not embed the assets, but provide the embedding API. Contents will still be loaded from disk.
  -dev
        Similar to debug, but does not emit absolute paths. Expects a rootDir variable to already exist in the generated code's package.
  -ignore value
        Regex pattern to ignore
  -mode uint
        Optional file mode override for all files.
  -modtime int
        Optional modification unix timestamp override for all files.
  -nocompress
        Assets will *not* be GZIP compressed when this flag is specified.
  -nomemcopy
        Use a .rodata hack to get rid of unnecessary memcopies. Refer to the documentation to see what implications this carries.
  -nometadata
        Assets will not preserve size, mode, and modtime info.
  -o string
        Optional name of the output file to be generated. (default "./bindata.go")
  -pkg string
        Package name to use in the generated code. (default "main")
  -prefix string
        Optional path prefix to strip off asset names.
  -tags string
        Optional set of build tags to include.
  -version
        Displays version information.
```

This is on 
```
4.8.0-49-generic #52~16.04.1-Ubuntu
```

I tried looking through spots in the code history for `github.com/jteeuwen/go-bindata` but couldn't find anywhere that a `md5checksum` flag was defined (looking with grep for `md5checksum`)

## Version
master and v0.9.0

## Expected Behavior
Build should not fail. Executing `make` should build the binary.

## Actual Behavior
Build fails

## Steps To Reproduce

1. Clone the repository
1. Use golang 1.8
1. Run make

## Logs
```
crallen@coreos-dev:~/src/rexray$ make
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
make deps
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   608    0   608    0     0   1498      0 --:--:-- --:--:-- --:--:--  1497
100 3690k  100 3690k    0     0   992k      0  0:00:03  0:00:03 --:--:-- 1530k
GOOS=linux GOARCH=amd64 go install _/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata
can't load package: package _/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata: cannot find package "_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata" in any of:
        /home/crallen/.go/go-1.8/src/_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata (from $GOROOT)
        /home/crallen/.go/src/src/_/home/crallen/src/rexray/vendor/github.com/jteeuwen/go-bindata/go-bindata (from $GOPATH)
Makefile:529: recipe for target '/home/crallen/.go/src/bin/go-bindata' failed
make[1]: *** [/home/crallen/.go/src/bin/go-bindata] Error 1
Makefile:80: recipe for target 'all' failed
make: *** [all] Error 2
```

```
crallen@coreos-dev:~/src/rexray$ go get -u github.com/jteeuwen/go-bindata/...
crallen@coreos-dev:~/src/rexray$
```

```
crallen@coreos-dev:~/src/rexray$ make
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
make deps
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
go get -u github.com/alecthomas/gometalinter
/home/crallen/.go/src/bin/gometalinter --install --update && touch .gometalinter.tools.d
Installing:
  goconst
  gocyclo
  goimports
  golint
  staticcheck
  unconvert
  deadcode
  gotype
  unparam
  aligncheck
  gas
  gosimple
  lll
  misspell
  safesql
  varcheck
  dupl
  errcheck
  ineffassign
  interfacer
  structcheck
  unused
make build
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
make build-libstorage
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
cd vendor/github.com/codedellemc/libstorage && \
        BUILD_TAGS="gofig pflag libstorage_integration_driver_linux scripts_generated" make api/server/executors/executors_generated.go && \
        cd -
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:828: warning: overriding recipe for target 'true'
Makefile:828: warning: ignoring old recipe for target 'true'
Makefile:828: warning: overriding recipe for target 'true-clean'
Makefile:828: warning: ignoring old recipe for target 'true-clean'
Makefile:891: warning: overriding recipe for target 'api/server/executors/bin/'
Makefile:890: warning: ignoring old recipe for target 'api/server/executors/bin/'
/home/crallen/.go/src/bin/go-bindata -md5checksum -pkg executors -prefix api/server/executors/bin -o api/server/executors/executors_generated.go api/server/executors/bin/...
flag provided but not defined: -md5checksum
Usage: /home/crallen/.go/src/bin/go-bindata [options] <input directories>

  -debug
        Do not embed the assets, but provide the embedding API. Contents will still be loaded from disk.
  -dev
        Similar to debug, but does not emit absolute paths. Expects a rootDir variable to already exist in the generated code's package.
  -ignore value
        Regex pattern to ignore
  -mode uint
        Optional file mode override for all files.
  -modtime int
        Optional modification unix timestamp override for all files.
  -nocompress
        Assets will *not* be GZIP compressed when this flag is specified.
  -nomemcopy
        Use a .rodata hack to get rid of unnecessary memcopies. Refer to the documentation to see what implications this carries.
  -nometadata
        Assets will not preserve size, mode, and modtime info.
  -o string
        Optional name of the output file to be generated. (default "./bindata.go")
  -pkg string
        Package name to use in the generated code. (default "main")
  -prefix string
        Optional path prefix to strip off asset names.
  -tags string
        Optional set of build tags to include.
  -version
        Displays version information.
Makefile:895: recipe for target 'api/server/executors/executors_generated.go' failed
make[3]: *** [api/server/executors/executors_generated.go] Error 2
Makefile:862: recipe for target 'vendor/github.com/codedellemc/libstorage/api/server/executors/executors_generated.go' failed
make[2]: *** [vendor/github.com/codedellemc/libstorage/api/server/executors/executors_generated.go] Error 2
Makefile:1263: recipe for target 'build' failed
make[1]: *** [build] Error 2
Makefile:80: recipe for target 'all' failed
make: *** [all] Error 2

```

**Comment by akutz** [comment]
Look at the Makefile. It uses Glide to pull my fork of GBD. You probably already have the binary in place, causing a target update issue. I opened a PR for that flag over a year ago at https://github.com/jteeuwen/go-bindata/pull/128.

**Comment by drcrallen** [comment]
@akutz thanks for the info. Tracing through the file it seems to only pull in your fork if you are building the agent?

trying to build the agent explicitly also fails with

```
crallen@coreos-dev:~/src/rexray$ git status
HEAD detached at v0.9.0
nothing to commit, working directory clean
crallen@coreos-dev:~/src/rexray$ make build-agent
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
REXRAY_BUILD_TYPE=agent make build
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
make build-libstorage
Makefile:823: warning: overriding recipe for target 'true'
Makefile:823: warning: ignoring old recipe for target 'true'
Makefile:823: warning: overriding recipe for target 'true-clean'
Makefile:823: warning: ignoring old recipe for target 'true-clean'
cd vendor/github.com/codedellemc/libstorage && \
        BUILD_TAGS="gofig pflag libstorage_integration_driver_linux scripts_generated rexray_build_type_agent" make api/api_generated.go && \
        cd -
/bin/bash: line 0: cd: vendor/github.com/codedellemc/libstorage: No such file or directory
Makefile:862: recipe for target 'vendor/github.com/codedellemc/libstorage/api/api_generated.go' failed
make[2]: *** [vendor/github.com/codedellemc/libstorage/api/api_generated.go] Error 1
Makefile:1263: recipe for target 'build' failed
make[1]: *** [build] Error 2
Makefile:1274: recipe for target 'build-agent' failed
make: *** [build-agent] Error 2
```

**Comment by akutz** [comment]
No, the controller or standalone. Remember, it all works on Travis-CI. https://travis-ci.org/codedellemc/rexray

**Comment by akutz** [comment]
Your logs show a bad build path and a bad dev env.

**Comment by drcrallen** [comment]
https://rexray.readthedocs.io/en/stable/dev-guide/build-reference/ simply does not work. This ended up working:

```
$ go get github.com/codedellemc/rexray
$ cd $GOPATH/src/github.com/codedellemc/rexray; git checkout v0.9.0 && make
```

**Comment by drcrallen** [comment]
```
The rexray binary is 26MB and located at:

  /home/crallen/.go/src/bin/rexray
```

**Comment by drcrallen** [comment]
`  /home/crallen/.go/src` is my `$GOPATH`

**Comment by akutz** [comment]
You're either not reading documentation closely or ignoring things on purpose:

1) The doc site works fine.
2) You're ignoring the *Note* on that doc site that says the command only works for Docker builds.
3) Of course you have to clone the code to the proper package path; that's how Go works.

RTFM

**Comment by akutz** [comment]
**Docker Builds**
![image](https://cloud.githubusercontent.com/assets/101085/26031644/da52003a-3840-11e7-8f38-d7c620ecc16a.png)

**Comment by akutz** [comment]
You were trying to build on a system without following the directions and apparently not knowing where to clone Go sources. Yes, you can clone them in *any* directory when building with Docker, but not with Go. You were building inside CoreOS, a famously poor Dev Env, in the wrong path, with a Makefile that was failing because you had none of the dev tools (or Docker) installed.

**Comment by akutz** [comment]
Hi @drcrallen,

You may also wish to see #846 and this [gist](https://gist.github.com/akutz/15e7d6610196fb885a015646824112ad).

**Comment by akutz** [comment]
Hi @drcrallen,

I agree that the note about Docker should indicate if Docker is not detected then the above command *will* fail. However, the Advanced Builds section lists Go as a requirement for non-Docker builds and lists the other build/dev reqs as well. If you, or anyone, is going to build Go source code, it's not up to the project to ensure that a person knows how/where to do so. 

I get that you were following an instruction *very* literally, but there is enough information on that page for devs to build REX-Ray without (many) issues if they know how to use Go.

**Comment by akutz** [comment]
Hi @drcrallen,

This [link](https://golang.org/doc/code.html) discusses Go workspaces and where sources *must* exist in order to be built. The purpose of the Docker-enabled build is to side-step the arcane and annoying path structure requirements.

**Comment by akutz** [comment]
Hi @drcrallen,

I apologize for being rude up above. I was awake only because I'm dealing with a child who won't go back to sleep. I should have never replied as grumpy as I am at the moment.

Let me explain the workflow as you encountered it:

1. The build docs illustrate the Basic Build that required Docker (as it's documented on that page you linked). 
2. You don't have Docker, so the Makefile then fails gracefully to using Go, if you have it.
3. The issue is that Go-based builds require that sources are cloned into the appropriate package location (workspace) inside the `GOPATH`. 
4. This is where the breakdown occurs.

It's my personal thought on the matter that by declaring Docker as a requirement for Basic Builds and Go as a requirement for Advanced Builds, a Go dev will know and understand that while Docker-based builds support out-of-path-structure builds, Go-based builds will not.

You're obviously not a beginner to Software Engineering, and so that threw me. I did not realize that you are (apparently) somewhat new to Go. If I'm wrong about that, I (again) apologize. The fact that you correctly use formatted Markdown alone was a shock to me as I struggle daily to get so many people to paste monospaced information correctly :)

The litany of Makefile errors you encountered were due to the fact that your build was using a Go-based solution in an invalid path. You were attempting to build the source code and thus I had assumed you knew about Go workspace requirements. Mea culpa.

To that end, I will be adding an **additional** warning to the Dev Docs that notes if you elect to clone the sources outside the standard path structure and your Make attempt doesn't use Docker (because it's not installed) and falls back to Go (because it *is* installed, but again, using the wrong path structure), that the build will fail.

I guess I never considered someone would have Go installed and **not** clone the sources inside the required path structure. The fact that you did, but were not using Docker to drive the build **and** had Go installed was a recipe for many bytes written to file descriptor 2.

**Comment by drcrallen** [comment]
I'm accustomed to various build environments but Go is a relatively new one for me that is true. The number three in your prior comment is the part I failed to grasp until you mentioned the build environment looked weird. Sorry for the late night noise and hopefully you guys get some sleeps soon.

**Comment by akutz** [comment]
Hi @drcrallen,

Holy crap I made a hugely embarrassing typo above:

> You're obviously **new** to Software Engineering, and so the situation threw me for a loop at first. What I failed to grasp was that you seem to be somewhat new to Go. If I'm wrong about that, I (again) apologize.

I edited that comment a few times, and at one point must have removed the **not**. I meant to say you're obviously **NOT** new to Software Engineering, but new to Go.

Damn. I am very sorry (yet again) for that unintentional slight. That's why I was originally confused by your trouble, because you clearly **do** know what you're doing with regards to tools, code, etc., and I only realized later that you didn't know about `GOPATH` and package structure.

Again, I'm an idiot, and I'm sorry. Glad you got it working.
